# Homework 10 - 07.01.22 - Nina Braunmiller - k11923286

## Task 1

In [1]:
ar1 <- matrix(c(1,1,0), 1)[rep(1,7), ]
ar2 <- matrix(c(1,0,1), 1)[rep(1,9), ] 
A <- rbind(ar1, ar2)
A
Y <- c(0.37, 41.86, 0.27, 12.21, 12.92, 4.25, 8.35, 0, 0, 0, 6.4, 0, 0.014, 0, 0.569, 1.59)

[,1] [,2] [,3]
 [1,] 1    1    0   
 [2,] 1    1    0   
 [3,] 1    1    0   
 [4,] 1    1    0   
 [5,] 1    1    0   
 [6,] 1    1    0   
 [7,] 1    1    0   
 [8,] 1    0    1   
 [9,] 1    0    1   
[10,] 1    0    1   
[11,] 1    0    1   
[12,] 1    0    1   
[13,] 1    0    1   
[14,] 1    0    1   
[15,] 1    0    1   
[16,] 1    0    1

### Compute the generalized inverse of A.T A

In [2]:
#install.packages('MASS')
library(MASS)

In [3]:
B = ginv(t(A)%*%A, tol=sqrt(.Machine$double.eps))
B

[,1]        [,2]        [,3]        
[1,] 0.028218695  0.01940035  0.008818342
[2,] 0.019400353  0.07583774 -0.056437390
[3,] 0.008818342 -0.05643739  0.065255732

### Compute the estimator and its variance for β1 + β2. Give an interpretation of the result of the estimator in the context. Is Var(βˆ1 + βˆ2) unique?

In [4]:
library(Matrix)

In [5]:
C = matrix(data=c(1,1,0), nrow=1, byrow=T)


In [6]:
OLS1 = C %*% B %*% t(A)
OLS1 # estimated coefficients for data vector y.


[,1]      [,2]      [,3]      [,4]      [,5]      [,6]      [,7]     
[1,] 0.1428571 0.1428571 0.1428571 0.1428571 0.1428571 0.1428571 0.1428571
     [,8]          [,9]          [,10]         [,11]         [,12]        
[1,] -6.938894e-18 -6.938894e-18 -6.938894e-18 -6.938894e-18 -6.938894e-18
     [,13]         [,14]         [,15]         [,16]        
[1,] -6.938894e-18 -6.938894e-18 -6.938894e-18 -6.938894e-18


When estimating y with $\beta_1 + \beta_2$ we get this coefficent-$\hat{y}$-vector. As you can see for the first seven entries we get the same values because the first seven rows in the given matrix A are the same, they get the same weight: $\frac{1}{numberRow_{1/.../7}} = \frac{1}{7} = 0.14...$<br>
Same counts for the last nine rows of A and the coefficent-$\hat{y}$-vector entries which are rounded to 0. That makes sense because we have: $1 \cdot \beta_1 + 0 \cdot \beta_2 = \hat{y}_{8/.../16}$. We need to ignore $\beta_3$ here because it is not considered in $\beta_1 + \beta_2$. Therefore, $\beta_1$ can't be estimated alone because the A matrix has rank 2 and the $\beta_2$ gets weight 0 here. It follows that, $y_{8/.../16}$ get weight of 0 here.


In [7]:
var = C %*% B %*% t(C)
var

[,1]     
[1,] 0.1428571

Its variance is unique as noted in the lecture. 


### Compute the estimator and its variance for β1 + β3. Give an interpretation of the result of the estimator in the context.

In [8]:
C2 = matrix(data=c(1,0,1), nrow=1, byrow=T)


In [9]:
OLS2 = C2 %*% B %*% t(A)
OLS2 # estimated coefficients for data vector y.


[,1]         [,2]         [,3]         [,4]         [,5]        
[1,] 2.775558e-17 2.775558e-17 2.775558e-17 2.775558e-17 2.775558e-17
     [,6]         [,7]         [,8]      [,9]      [,10]     [,11]    
[1,] 2.775558e-17 2.775558e-17 0.1111111 0.1111111 0.1111111 0.1111111
     [,12]     [,13]     [,14]     [,15]     [,16]    
[1,] 0.1111111 0.1111111 0.1111111 0.1111111 0.1111111

Same interpretation and explanation as above but here reversed order because A[row=8/.../16] = [1,0,1] such that $\beta_1$ and $ \beta_3$ are considered and $\beta_2$ ignored. Because all the latter described rows of A are the same, the coefficients for y have all the same weights $\frac{1}{numberRow_{8/.../16})} = \frac{1}{9} = 0.11...$<br>.
A[row=1/.../7] can't be estimated by the current parameter sum. Therefore, here we observe a weight of 0 for the y-entries. 

In [10]:
var = C2 %*% B %*% t(C2)
var


[,1]     
[1,] 0.1111111

## Task 2

Let's go through my thinking procedure. On the first glance the formula seems to fit but the assumption of homoscedasticity of errors changed my mind.<br>
Given: $\sigma^2 ?= \frac{\sum_{i=1}^n{(y_i - \hat{\beta_1}- \hat{\beta_i})^2}}{n-rank(X)} = 
\frac{\sum_{i=1}^n{(y_i - (\hat{\beta_1}+ \hat{\beta_i}))^2}}{3-rank(A)} =
\frac{(y_2 - ( \hat{\beta_1} +\hat{\beta_2}))^2 + (y_3 -( \hat{\beta_1}+ \hat{\beta_3}))^2}{16-2}$ 

So far so good. $y_2$ refers to all y-values whos x-values got a weight > 0 by $(\hat{\beta_1}+ \hat{\beta_2})$. Same counts for $y_3$. <br>
We have all values of the variables and ols in $(y_2 - (\hat{\beta_1}+ \hat{\beta_2}))^2 + (y_3 - (\hat{\beta_1}+ \hat{\beta_3}))^2$. It could be easy to caculate a value from it. Furthermore, the lecture slides also include X in this formula like this: $(y_2 - X(\hat{\beta_1}+ \hat{\beta_2}))^2 + (y_3 - X(\hat{\beta_1}+ \hat{\beta_3}))^2$. X is our matrix A and only contains 1- and 0-values. So, this is fine because it nothing changes at our given formula. <br>

What is the problem then? We have the assumption of homoscedasticity of errors. That means that residuals have always about the same size (no significant difference between them) related to the model. For the error matrix we have as small example: $\epsilon =\sigma^2  \left[  \begin{matrix}
  1 & 0\\
  0 & 1
  \end{matrix} \right]$.<br> Also we have: $Var(\epsilon_i)=x_i\sigma^2$.<br>
$\sigma^2$ has to stay always the same! But we have two different ols models to explain the same y. $(\hat{\beta_1}+ \hat{\beta_2})$ related to seven rows of A and $(\hat{\beta_1}+ \hat{\beta_3})$ to nine rows such that they gave different weights to the affected lines in estimated coefficients vector y which are also their variances. 



In [11]:
(Y[1:7] - OLS1[1:7])^2 # (y_2 - (\hat{\beta_1}+ \hat{\beta_2}))^2

[1] 5.159388e-02 1.740320e+03 1.616531e-02 1.456159e+02 1.632554e+02
[6] 1.686862e+01 6.735719e+01

In [12]:
(Y[8:16] - OLS2[8:16])^2 #(y_3 - (\hat{\beta_1}+ \hat{\beta_3}))^2

[1]  0.012345679  0.012345679  0.012345679 39.550123457  0.012345679
[6]  0.009430568  0.012345679  0.209662235  2.187112346

When using the formula even the lengths differ. Also the values are obviously different.<br><br>
Answer: no

## Task 3

In [13]:
library(nortest)

Compute the residuals. A is our design matrix. With:
\begin{align*}
Y_{1/.../7} &= X_{1/.../7 }(\beta_1 , \beta_2, \beta_3)^T + \epsilon \\
&= \left[  \begin{matrix}
  1 \beta_1 + 1 \beta_2 + 0 \beta_3\\
  ...
  \end{matrix} \right] + \epsilon \\
&=  \left[  \begin{matrix}
  \beta_1 + \beta_2\\
  ...
  \end{matrix} \right] + \epsilon \\
&=  \left[  \begin{matrix}
  0.1428571\\
  ...
  \end{matrix} \right] + \epsilon \\ 
  \\
  \\
\epsilon &= Y_{1/.../7} - \left[  \begin{matrix}
  0.1428571\\
  ...
  \end{matrix} \right] \\ 
\end{align*}

In [14]:
Y1 <- matrix(Y[1:7])
resi1 <- Y1 -  matrix(OLS1[1:7])
resi1

[,1]      
[1,]  0.2271429
[2,] 41.7171429
[3,]  0.1271429
[4,] 12.0671429
[5,] 12.7771429
[6,]  4.1071429
[7,]  8.2071429

In [15]:
Y2 <- matrix(Y[8:16])
resi2 <- Y2 -  matrix(OLS2[8:16])
resi2

[,1]       
 [1,] -0.11111111
 [2,] -0.11111111
 [3,] -0.11111111
 [4,]  6.28888889
 [5,] -0.11111111
 [6,] -0.09711111
 [7,] -0.11111111
 [8,]  0.45788889
 [9,]  1.47888889

In [16]:
residuals_ <- rbind(resi1,resi2)

In [17]:
# 1. Anderson_Darling test for normality:
p1 <- ad.test(residuals_)$p.value
# 2. Cramer-von Mises test for normality:
p2 <- cvm.test(residuals_)$p.value
# 3. Lilliefors (Kolmogorov-Smirnov) test for normality:
p3 <- lillie.test(residuals_)$p.value
# 4. Pearson chi-square test for normality:
p4 <- pearson.test(residuals_, n.classes=length(residuals_))$p.value
# 5. Shapiro_Francia test for normality:
p5 <- sf.test(residuals_)$p.value
p_average <- (p1+p2+p3+p4+p5)/5

In [18]:
p1
p2
p3
p4
p5
p_average

[1] 1.781308e-06

[1] 9.322175e-06

[1] 0.0003905466

[1] 6.543016e-06

[1] 2.414225e-05

[1] 8.646708e-05

The tests refer to the composite hypothesis of normality of the inputted x (here residuals).
All p < significance_level = 0.05 which means that we can reject the nullhypothesis that the residuals come from a normal distribution. We can integrate that knowledge
when creating a second better fitting model.

We can also look at the Komogorov-Smirnov test with which we can look if there is a difference between our residuals and a normal distribution.

In [19]:
residuals_removed_duplicates <- residuals_[!duplicated(residuals_), ] # remove the duplicates, else the 
# test won't work
k1<-ks.test(residuals_removed_duplicates,"pnorm") # test if samples come from normal distribution
k1


	One-sample Kolmogorov-Smirnov test

data:  residuals_removed_duplicates
D = 0.51375, p-value = 0.001822
alternative hypothesis: two-sided


p < 0.01 < 0.05. When this happens we can reject the null-hypothesis which says that there is NO difference between a subsample set and a normal distribution. So, indeed we assume that there is a difference between our residuals and normal distribution.

The results give us a reason to assume the hyothesis that the residuals aren't normally distributed.